### PR TITLE
[IMP] website_sale: add hook to _get_checkout_steps for easier override

### PR DESCRIPTION
**Pull request by daonamutc1**
`_get_checkout_steps` didn't provide a friendly way to add steps in the
checkout flow. Developers had to reimplement part of the parent logic in 
their overrides.

This commit introduces `get_checkout_step_list`. This method will allow
an easier override of the checkout steps while keeping the logic to 
return a single step in the parent method.

### CODE REVIEW

**Comment by daonamutc1** [comment]
@mathieu-mgm can you review?

**Comment by daonamutc1** [comment]
@mano-odoo can you review?

**Comment by mano-odoo** [comment]
Hello @daonamutc1, I would be happy to review design related changes, but for this PR it would be better to ping @odoo/rd-ecommerce 😄

**Comment by daonamutc1** [comment]
> Hello @daonamutc1, I would be happy to review design related changes, but for this PR it would be better to ping @odoo/rd-ecommerce 😄

Thank you, it's really hard to find the exact author of this piece of code. Sorry for the inconvenience.

**Comment by daonamutc1** [comment]
@Feyensv Can you review

File: addons/website_sale/models/website.py [hunk 1]
```diff
@@ -564,32 +564,10 @@ def _display_partner_b2b_fields(self):
...
-                    'name': str,
-                    'current_href': str,
-                    'main_button': str,
-                    'main_button_href': str,
-                    'back_button': str,
-                    'back_button_href': str
-                }
-            ]
-        """
+    def _get_checkout_steps_list(self):
```
Comment (chevalierv) [inline addons/website_sale/models/website.py:567]: Please add a complete docstring
```suggestion
    def _get_checkout_step_list(self):
        """ Return an ordered list of steps according to the current template rendered.
      
        :rtype: list
        :return: A list with the following structure:
            [
                [xmlid],
                {
                    'name': str,
                    'current_href': str,
                    'main_button': str,
                    'main_button_href': str,
                    'back_button': str,
                    'back_button_href': str
                }
            ]
        """
```

File: addons/website_sale/models/website.py [hunk 2]
```diff
@@ -620,6 +598,33 @@ def _get_checkout_steps(self, current_step=None):
...
+                {
+                    'name': str,
+                    'current_href': str,
+                    'main_button': str,
+                    'main_button_href': str,
+                    'back_button': str,
+                    'back_button_href': str
+                }
+            ]
+        """
```
Comment (chevalierv) [inline addons/website_sale/models/website.py:624]: Don't forget to adapt the docstring
```suggestion
        """ Return an ordered list of steps according to the current template rendered.
        If `current_step` is provided, returns only the corresponding step.
        Note: self.ensure_one()
        :param str current_step: The xmlid of the current step, defaults to None.
        :rtype: list
        :return: A list containing the steps generated by :meth:`_get_checkout_step_list`.
        """
```

File: addons/website_sale/models/website.py [hunk 3]
```diff
@@ -564,32 +564,10 @@ def _display_partner_b2b_fields(self):
...
-                    'back_button': str,
-                    'back_button_href': str
-                }
-            ]
-        """
+    def _get_checkout_steps_list(self):
         self.ensure_one()
         is_extra_step_active = self.viewref('website_sale.extra_info').active
         redirect_to_sign_in = self.account_on_checkout == 'mandatory' and self.is_public_user()
-
```
Comment (chevalierv) [inline addons/website_sale/models/website.py:592]: Why remove this blank line?

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @daonamutc1 

As you can see in the PR, C3POdoo has already ping the righteous team to review your PR. There is no need to ping everyone on it.

It is possible to add steps to this method, this method is already overridden in our enterprise repository. You can get all the steps by calling the parent like this:
`parent_steps = super()._get_checkout_steps(current_step=None)`
But I do agree that it is not a perfect solution as you'll need to look for `current_step` in your override too.

I've left some comments about your PR, please apply them so I can merge the PR afterward.

Also, please change your commit title to something a little bit shorter so everything fits on one line. Here is a suggestion: `add hook to _get_checkout_steps for easier override` 

Have a nice day!

File: addons/website_sale/models/website.py [hunk 4]
```diff
@@ -620,6 +615,19 @@ def _get_checkout_steps(self, current_step=None):
...
+        return steps
+
+    def _get_checkout_steps(self, current_step=None):
+        """ Return an ordered list of steps according to the current template rendered.
+	        If `current_step` is provided, returns only the corresponding step.
+	        Note: self.ensure_one()
+	        :param str current_step: The xmlid of the current step, defaults to None.
+	        :rtype: list
+	        :return: A list containing the steps generated by :meth:`_get_checkout_step_list`.
+	    """
```
Comment (chevalierv) [inline addons/website_sale/models/website.py:627]: :eyes: 
```suggestion
	      If `current_step` is provided, returns only the corresponding step.
	      Note: self.ensure_one()
	      :param str current_step: The xmlid of the current step, defaults to None.
	      :rtype: list
	      :return: A list containing the steps generated by :meth:`_get_checkout_step_list`.
	  """
```

**Comment by daonamutc1** [comment]
> Hello @daonamutc1
> 
> As you can see in the PR, C3POdoo has already ping the righteous team to review your PR. There is no need to ping everyone on it.
> 
> It is possible to add steps to this method, this method is already overridden in our enterprise repository. You can get all the steps by calling the parent like this: `parent_steps = super()._get_checkout_steps(current_step=None)` But I do agree that it is not a perfect solution as you'll need to look for `current_step` in your override too.
> 
> I've left some comments about your PR, please apply them so I can merge the PR afterward.
> 
> Also, please change your commit title to something a little bit shorter so everything fits on one line. Here is a suggestion: `add hook to _get_checkout_steps for easier override`
> 
> Have a nice day!

I tried `parent_steps = super()._get_checkout_steps(current_step=None)`. However, after adding a new step, `next(step for step in steps if current_step in step[0])[1]` will raise an error because the new step is not in `steps`.

File: addons/website_sale/models/website.py [hunk 5]
```diff
@@ -620,6 +615,19 @@ def _get_checkout_steps(self, current_step=None):
...
+        """ Return an ordered list of steps according to the current template rendered.
+        If `current_step` is provided, returns only the corresponding step.
+        Note: self.ensure_one()
+        :param str current_step: The xmlid of the current step, defaults to None.
+        :rtype: list
+        :return: A list containing the steps generated by :meth:`_get_checkout_step_list`.
+        """
+        self.ensure_one()
+
+        steps = self._get_checkout_steps_list()
```
Comment (chevalierv) [inline addons/website_sale/models/website.py:630]:
```suggestion
        steps = self._get_checkout_step_list()
```

**Review by chevalierv — CHANGES_REQUESTED** [review]

**Comment by chevalierv** [comment]
@robodoo r+ squash

**Pull request closed without merging**